### PR TITLE
Recurring sales order

### PR DESCRIPTION
**Pull request by interhawk**
Database changes to enable creating recurring sales orders

### CODE REVIEW

File: foundation-database/public/functions/copysoheader.sql [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+  END IF;
+
+  INSERT INTO charass
+        (charass_target_type, charass_target_id,
+         charass_char_id, charass_value)
+  SELECT charass_target_type, _soheadid,
+         charass_char_id, charass_value
+    FROM charass
+   WHERE ((charass_target_type='SO')
+     AND  (charass_target_id=pSoheadid));
```
Comment (andersdd) [inline foundation-database/public/functions/copysoheader.sql]: Don't know the answer to this (@gilmoskowitz) but should we also copy the document assignments in `docass`?

File: foundation-database/public/patches/populate_recur.sql [hunk 2]
```diff
@@ -0,0 +1,6 @@
+INSERT INTO recurtype (recurtype_type, recurtype_table, recurtype_donecheck, 
+                       recurtype_schedcol, recurtype_limit, recurtype_copyfunc, 
+                       recurtype_copyargs, recurtype_delfunc) 
+               VALUES ('SALESORDER', 'cohead', 'isRecurSODone(cohead_id)', 'cohead_orderdate',
```
Comment (andersdd) [inline foundation-database/public/patches/populate_recur.sql]: Should `'SALESORDER'` actually be `'S'` ?

File: foundation-database/public/trigger_functions/cohead.sql [hunk 3]
```diff
@@ -651,16 +651,41 @@ CREATE TRIGGER coheadBeforeDeleteTrigger
...
   DELETE FROM charass
   WHERE charass_target_type = 'SO'
     AND charass_target_id = OLD.cohead_id;
 
+  --recurrence cleanup
+  SELECT recur_id INTO _recurid
+    FROM recur
+   WHERE ((recur_parent_id = OLD.cohead_id)
+     AND  (recur_parent_type = "SALESORDER"));
+  IF (_recurid IS NOT NULL) THEN -- The deleted SO is the parent of a recurrence series
```
Comment (andersdd) [inline foundation-database/public/trigger_functions/cohead.sql]: `SALESORDER` should be switched to 'S'.  Also watch the xTuple coding standards when it comes to excessive brackets in WHERE clauses

File: foundation-database/public/functions/copysoheader.sql (hunk 1 above)
Comment (interhawk) [reply foundation-database/public/functions/copysoheader.sql]: the new copysoheader procedure is simply a portion of the existing copyso procedure that was pulled out into a separate function to enable copying a sales order without copying line items. Copying document assignments may well be a good thing to consider, but  I think outside the scope of this pr maybe?

File: foundation-database/public/patches/populate_recur.sql (hunk 2 above)
Comment (interhawk) [reply foundation-database/public/patches/populate_recur.sql]: Good catch, done.

File: foundation-database/public/trigger_functions/cohead.sql (hunk 3 above)
Comment (interhawk) [reply foundation-database/public/trigger_functions/cohead.sql]: implemented

File: foundation-database/public/functions/copysoheader.sql (hunk 1 above)
Comment (gilmoskowitz) [reply foundation-database/public/functions/copysoheader.sql]: I agree with @interhawk — out of scope. Copying `docass` is fraught with danger, as some documents might contain information specific to the original order (e.g. signed contracts) while others might be boilerplate.

**Pull request merged**